aspnet/mvc/controllers/testing.rst

```diff
@@ -1,8 +1,173 @@
+  :language: c#
+  :lines: 12-61
+  :dedent: 4
+  :emphasize-lines: 3-4,9,17-18,21,30
+
+The first test confirms when ``ModelState`` is not valid, the same ``ViewResult`` is returned as for a ``GET`` request. Note that the test doesn't attempt to pass in an invalid model. That wouldn't work anyway since model binding isn't running - we're just calling the method directly. However, we're not trying to test model binding - we're only testing what our code in the action method does. The simplest approach is to add an error to ``ModelState``.
+
+The second test verifies that when ``ModelState`` is valid, a new ``BrainstormSession`` is added (via the repository), and the method returns a ``RedirectToActionResult`` with the expected properties. Mocked calls that aren't called are normally ignored, but calling ``Verifiable`` at the end of the setup call allows it to be verified in the test. This is done with the call to ``mockRepo.Verify``.
+
+.. note:: The Moq library used in this sample makes it easy to mix verifiable, or "strict", mocks with non-verifiable mocks (also called "loose" mocks or stubs). Learn more about `customizing Mock behavior with Moq <https://github.com/Moq/moq4/wiki/Quickstart#customizing-mock-behavior>`_.
+
+Another controller in the app displays information related to a particular brainstorming session. It includes some logic to deal with invalid id values:
+
+.. literalinclude:: testing/sample/TestingControllersSample/src/TestingControllersSample/Controllers/SessionController.cs
+  :language: c#
+  :emphasize-lines: 16,20,25,33
+
+The controller action has three cases to test, one for each ``return`` statement:
+
+.. literalinclude:: testing/sample/TestingControllersSample/tests/TestingControllerSample.Tests/UnitTests/SessionControllerIndex.cs
+  :language: c#
+  :emphasize-lines: 16,26,39
+
+The app exposes functionality as a web API (a list of ideas associated with a brainstorming session and a method for adding new ideas to a session):
+
+.. _ideas-controller:
+
+.. literalinclude:: testing/sample/TestingControllersSample/src/TestingControllersSample/Api/IdeasController.cs
+  :language: c#
+  :emphasize-lines: 20-22,27,29-35,49-51,55,60,70
+
+The ``ForSession`` method returns a list of anonymous types, with property names camel cased to match JavaScript conventions. Avoid returning your business domain entities directly via API calls, since frequently they include more data than the API client requires, and they unnecessarily couple your app's internal domain model with the API you expose externally. You can define strongly typed data-transfer objects (DTOs), or just return an anonymous type as this method does. Mapping between domain entities and the types you will return over the wire can be done manually (using a LINQ ``Select`` as shown here) or using a library like `AutoMapper <https://github.com/AutoMapper/AutoMapper>`_
```

Comment:
When I first read this it felt a bit off topic. After reading the section that follows I see that there is a testing concern,  but it might help to address the testing issues associated response types more directly.